src/main/java/io/vertx/core/impl/VertxImpl.java

```diff
@@ -668,10 +683,18 @@ public void undeploy(String deploymentID) {
 
   @Override
   public void undeploy(String deploymentID, Handler<AsyncResult<Void>> completionHandler) {
-    if (haManager() != null && haManager().isEnabled()) {
-      haManager().removeFromHA(deploymentID);
-    }
-    deploymentManager.undeployVerticle(deploymentID, completionHandler);
+    this.executeBlocking(fut -> {
+      if (haManager() != null && haManager().isEnabled()) {
```

Comment:
move the condition outside of the executeBlocking block so we don't execute the blocking task if there is no need to be. In addition you can capture the haManager in a local variable.